### DISCUSSION TITLE
wrong result for morphology opening

### DISCUSSION THREAD

**Discussion by 24hours** [kornia/kornia#1188] [Questions] [answered]
## To Reproduce

```python
import torch
import numpy as np
from scipy.ndimage.morphology import grey_opening
from kornia.morphology import opening
from skimage.morphology import opening as sk_opening
kernel = np.array([[0, 0, 0, 0, 1, 0, 0, 0, 0],
                   [0, 0, 1, 1, 1, 1, 1, 0, 0],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [1, 1, 1, 1, 1, 1, 1, 1, 1],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [0, 0, 1, 1, 1, 1, 1, 0, 0],
                   [0, 0, 0, 0, 1, 0, 0, 0, 0]], dtype=np.float32)

input = np.zeros([1, 1, 128, 128])
output_scipy = grey_opening(input[0, 0, :, :], structure=kernel)
output_sk = sk_opening(input[0, 0, :, :], kernel)
output_kornia = opening(torch.tensor(input), torch.tensor(kernel))
print(np.all(output_scipy == output_sk))
print(np.all(output_sk == output_kornia))
print(np.all(output_scipy == output_kornia))

```

this bug present on master branch, `opening` is not yet in any release version

**Comment by edgarriba** [comment]
@24hours thanks for reporting. I would compare with a non zeros input. /cc @Manza12

**Comment by Manza12** [comment]
This bug is produced by the padding of the border. I will implement a custom border value for all the morphological operations as it is done in [nnMorpho](https://github.com/Manza12/nnMorpho). This way you would be able to fully control the border value. 

Also, comparing PyTorch tensors with numpy arrays won't work for np.all; last two lines should be 
```
print(np.all(output_sk == output_kornia.numpy()))
print(np.all(output_scipy == output_kornia.numpy()))
```
Nevertheless, I checked the images and it is true that they don't give the same result. When testing with `input = np.ones([1, 1, 128, 128])` they do, though

**Comment by edgarriba** [comment]
@Manza12 is this still relevant ?

**Comment by Manza12** [comment]
I think it is solved. But I don’t remeber well what was this about. I am going to check it and I Will answer properly.

**Comment by edgarriba** [comment]
@Manza12 awesome - feel free to close

**Comment by Manza12** [comment, marked as answer]
This issue is solved. Try:
```
import torch
import numpy as np
from scipy.ndimage.morphology import grey_opening
from kornia.morphology import opening
kernel = np.array([[0, 0, 0, 0, 1, 0, 0, 0, 0],
                   [0, 0, 1, 1, 1, 1, 1, 0, 0],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [1, 1, 1, 1, 1, 1, 1, 1, 1],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [0, 1, 1, 1, 1, 1, 1, 1, 0],
                   [0, 0, 1, 1, 1, 1, 1, 0, 0],
                   [0, 0, 0, 0, 1, 0, 0, 0, 0]], dtype=np.float32)

input = np.zeros([1, 1, 128, 128])
output_scipy = grey_opening(input[0, 0, :, :], structure=kernel)
output_kornia = opening(torch.tensor(input), torch.tensor(kernel), border_type='reflect')
print(np.all(output_scipy == output_kornia.numpy()))
```

Default mode for the border in Scipy is 'reflect'. You can change the border mode in kornia with the parameter `border_type` and, if 'constant', you can ajust the value with `border_value`. In addition, default mode for border in kornia.morphology is 'geodesic', which means that it only takes into account the pixels of the image for comouting the maximum/minimum.
